modules/build/src/main/scala/scala/build/options/BuildOptions.scala

```diff
@@ -273,81 +268,225 @@ final case class BuildOptions(
+
+    val moduleVersion: Either[ScalaVersionError, String] = {
+      import coursier._
+      def scalaNightly2Module: Module = mod"org.scala-lang:scala-library"
+      val res = finalCache.logger.use {
+        Versions(finalCache)
+          .withModule(scalaNightly2Module)
+          .withRepositories(Seq(coursier.Repositories.scalaIntegration))
+          .result()
+          .unsafeRun()(finalCache.ec)
+      }
+      res.versions.available.find(versionString == _) match {
+        case Some(vStr) => Right(vStr)
+        case None => Left(new NoValidScalaVersionFoundError(
+            res.versions.available,
+            latestSupportedStableVersions
+          ))
+      }
     }
 
-    val sv  = value(maybeSv)
-    val sbv = scalaBinaryVersion.getOrElse(ScalaVersion.binary(sv))
-    (sv, sbv)
+    val scalaVersion       = value(moduleVersion)
+    val scalaBinaryVersion = ScalaVersion.binary(scalaVersion)
+    (scalaVersion, scalaBinaryVersion)
+  }
+
+  private def turnScala3NightlyVersionArgIntoVersion(versionString: String)
+    : Either[BuildException, (String, String)] = either {
+    val moduleVersion: Either[ScalaVersionError, String] = {
+      import coursier._
+      def scala3 = mod"org.scala-lang:scala3-library_3"
+      val res = finalCache.logger.use {
+        Versions(finalCache)
+          .withModule(scala3)
+          .result()
+          .unsafeRun()(finalCache.ec)
+      }
+      res.versions.available.find(versionString == _) match {
+        case Some(vStr) => Right(vStr)
+        case None => Left(new NoValidScalaVersionFoundError(
+            res.versions.available,
+            latestSupportedStableVersions
+          ))
+      }
+    }
+
+    val scalaVersion       = value(moduleVersion)
+    val scalaBinaryVersion = ScalaVersion.binary(scalaVersion)
+    (scalaVersion, scalaBinaryVersion)
   }
 
   lazy val scalaParams: Either[BuildException, ScalaParameters] = either {
+    def isScala2Nightly(version: String): Boolean =
+      scala2NightlyRegex.unapplySeq(version).isDefined
+    def isScala3Nightly(version: String): Boolean =
+      version.startsWith("3") && version.endsWith("-NIGHTLY")
+
     val (scalaVersion, scalaBinaryVersion) =
-      value(computeScalaVersions(scalaOptions.scalaVersion, scalaOptions.scalaBinaryVersion))
+      scalaOptions.scalaVersion match {
+        case Some("3.nightly") => value(computeLatestScalaThreeNightlyVersions())
+        case Some("2.nightly") => value(computeLatestScalaTwoNightlyVersions())
+        case Some(versionString) if isScala3Nightly(versionString) =>
+          value(turnScala3NightlyVersionArgIntoVersion(versionString))
+        case Some(versionString) if isScala2Nightly(versionString) =>
+          value(turnScala2NightlyVersionArgToVersions(versionString))
+        case _ => value(turnScalaVersionArgToScalaVersions(
+            scalaOptions.scalaVersion,
+            scalaOptions.scalaBinaryVersion
+          ))
+      }
```

Comment:
The `value` can be factored here:
```suggestion
      value {
        scalaOptions.scalaVersion match {
          case Some("3.nightly") => computeLatestScalaThreeNightlyVersions()
          case Some("2.nightly") => computeLatestScalaTwoNightlyVersions()
          case Some(versionString) if isScala3Nightly(versionString) =>
            turnScala3NightlyVersionArgIntoVersion(versionString)
          case Some(versionString) if isScala2Nightly(versionString) =>
            turnScala2NightlyVersionArgToVersions(versionString)
          case _ => turnScalaVersionArgToScalaVersions(
              scalaOptions.scalaVersion,
              scalaOptions.scalaBinaryVersion
            )
        }
      }
```